algoliasearch/client.go

```diff
@@ -21,222 +19,187 @@ type Client struct {
 // parameters.
-func (c *Client) UpdateKeyWithParam(key string, params interface{}) (interface{}, error) {
-	return c.transport.request("PUT", "/1/keys/"+key, params, write)
+func (c *Client) UpdateKeyWithParam(key string, params map[string]interface{}) (res UpdateKeyRes, err error) {
+	if err = checkKey(params); err != nil {
+		return res, err
+	}
+
+	err = c.request(&res, "PUT", "/1/keys/"+key, params, write)
+	return res, err
 }
 
-// GetKey returns the characteristics of the API key named `key`.
-func (c *Client) GetKey(key string) (interface{}, error) {
-	return c.transport.request("GET", "/1/keys/"+key, nil, read)
+// GetKey returns the ACL and validity of the API key named `key`.
+func (c *Client) GetKey(key string) (res GetKeyRes, err error) {
+	err = c.request(&res, "GET", "/1/keys/"+key, nil, read)
+	return res, err
 }
 
 // DeleteKey deletes the API key named `key`.
-func (c *Client) DeleteKey(key string) (interface{}, error) {
-	return c.transport.request("DELETE", "/1/keys/"+key, nil, write)
+func (c *Client) DeleteKey(key string) (res DeleteKeyRes, err error) {
+	err = c.request(&res, "DELETE", "/1/keys/"+key, nil, write)
+	return res, err
 }
 
 // GetLogs retrieves the `length` latest logs, starting at `offset`. Logs can
 // be filtered by type via `logType` being either "query", "build" or "error".
-func (c *Client) GetLogs(offset, length int, logType string) (interface{}, error) {
-	body := map[string]interface{}{
-		"offset": offset,
-		"length": length,
-		"type":   logType,
+func (c *Client) GetLogs(params map[string]interface{}) (res GetLogsRes, err error) {
+	if err = checkGetLogs(params); err != nil {
+		return res, err
 	}
 
-	return c.transport.request("GET", "/1/logs", body, write)
+	err = c.request(&res, "GET", "/1/logs", params, write)
+	return res, err
 }
 
-// GenerateSecuredApiKey generates a public API key intended to restrict access
+// GenerateSecuredAPIKey generates a public API key intended to restrict access
 // to certain records.
 // This new key is built upon the existing key named `apiKey`. Tag filters
 // or query parameters used to restrict access to certain records are specified
 // via the `public` argument. A single `userToken` may be supplied, in order to
 // use rate limited access.
-func (c *Client) GenerateSecuredApiKey(apiKey string, public interface{}, userToken ...string) (string, error) {
-	if len(userToken) > 1 {
-		return "", errors.New("Too many parameters")
+func (c *Client) GenerateSecuredAPIKey(apiKey string, params map[string]interface{}) (string, error) {
+	if err := checkGenerateSecuredAPIKey(params); err != nil {
+		return "", nil
 	}
 
-	var userTokenStr string
-	var message string
-	if len(userToken) == 1 {
-		userTokenStr = userToken[0]
-	} else {
-		userTokenStr = ""
+	req := make(map[string]interface{})
+	for k, v := range params {
+		req[k] = v
 	}
+	req["tagFilters"] = strings.Join(params["tagFilters"].([]string), ",")
+	message := c.transport.EncodeParams(req)
```

Comment:
Why don't you simply use the original `params` map and change the `tagFilters` field?